executor/src/witgen/range_constraints.rs

```diff
@@ -1,86 +1,213 @@
 use super::symbolic_evaluator::SymbolicEvaluator;
 use super::util::try_to_simple_poly;
 use super::{Constraint, FixedData};
 
-/// Constraint on the bit values of a variable X.
-/// All bits smaller than min_bit have to be zero
-/// and all bits larger than max_bit have to be zero.
-/// The least significant bit is bit zero.
-#[derive(Clone, PartialEq, Eq)]
+/// Constraint on the values of a variable X.
+/// It does not have to be an interval.
+///
+/// Currently, we can represent interval ranges (both "wrapping" and "non-wrapping" ones)
+/// and bit masks. The actual constraint is the conjunction of the two.
+///
+/// Note that the same constraint can have multiple representations.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
 pub struct RangeConstraint<T: FieldElement> {
+    /// Bit-mask.
     mask: T::Integer,
+    /// Min-max inclusive range. Note that `max` can be smaller than `min`. In this case the range wraps.
+    min: T,
+    max: T,
 }
 
 impl<T: FieldElement> RangeConstraint<T> {
+    /// Constraint that allows no higher bits set than the one given
+    /// (counting from zero).
+    pub fn from_max_bit(max_bit: u64) -> Self {
+        Self::from_mask(mask_from_bits::<T>(max_bit + 1))
+    }
+
+    /// Constraint that forces several bits to be set to zero.
+    /// Semantics: x & mask == x
+    pub fn from_mask<M: Into<T::Integer>>(mask: M) -> Self {
+        let mask = mask.into();
+        let max = T::from(cmp::min(mask, (T::from(-1)).to_integer()));
+        Self {
+            mask,
+            min: T::zero(),
+            max,
+        }
+    }
+
+    /// Constraint that only allows this exact value.
+    pub fn from_value(value: T) -> Self {
+        Self {
+            mask: value.to_integer(),
+            min: value,
+            max: value,
+        }
+    }
+
+    /// Constraint that allows values from min to max.
+    /// If min <= max, this means min <= x && x <= max.
+    /// If min > max, this means min <= x || x <= max.
+    pub fn from_range(min: T, max: T) -> Self {
+        let mask = if min <= max {
+            mask_from_bits::<T>(max.to_integer().num_bits() as u64)
+        } else {
+            !T::Integer::from(0)
+        };
+        Self { mask, min, max }
+    }
+
+    /// Returns a bit mask. This might be drastically underfitted in case
+    /// the constraint is more resembling an interval.
+    /// Semantics: X & mask == X holds for all possible values of X.
     pub fn mask(&self) -> &T::Integer {
         &self.mask
     }
-}
 
-impl<T: FieldElement> RangeConstraint<T> {
-    pub fn from_max_bit(max_bit: u64) -> Self {
-        assert!(max_bit < 1024);
-        RangeConstraint {
-            mask: T::Integer::from(((1 << (max_bit + 1)) - 1) as u32),
-        }
+    /// Returns a min-max inclusive range. Note that `max` can be smaller than `min`. In this case the range wraps.
+    /// Semantics: X >= range().0 && X <= range().1 holds for all possible values of X.
```

Comment:
Not really `X >= range().0 && X <= range().1` for when min > max, right?

Reply:
Yes, will update. Its either `range().0 <= x && x <= range().1` or `range().0 <= x || x <= range().1`